server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -3205,8 +3211,9 @@ private EngineConfig newEngineConfig(LongSupplier globalCheckpointSupplier) {
 
         final List<ReferenceManager.RefreshListener> internalRefreshListener = new ArrayList<>();
         internalRefreshListener.add(new RefreshMetricUpdater(refreshMetric));
-        if (remoteStoreRefreshListener != null && shardRouting.primary()) {
-            internalRefreshListener.add(remoteStoreRefreshListener);
+        if (isRemoteStoreEnabled()) {
+            Directory remoteDirectory = ((FilterDirectory) ((FilterDirectory) remoteStore.directory()).getDelegate()).getDelegate();
+            internalRefreshListener.add(new RemoteStoreRefreshListener(store.directory(), remoteDirectory));
```

Comment:
Sorry I couldn't find where are we handling the reference counting for Store. The call to `directory() ` requires the Store to be Open(ref counter > 0) so this call would fail?

https://github.com/opensearch-project/OpenSearch/blob/3d4d5ca16547e891768729f8ead0b6941f5cc63f/server/src/main/java/org/opensearch/index/store/Store.java#L208-L211

Reply:
When store is created, default `refCount` is 1.

https://github.com/opensearch-project/OpenSearch/blob/3d4d5ca16547e891768729f8ead0b6941f5cc63f/libs/core/src/main/java/org/opensearch/common/util/concurrent/AbstractRefCounted.java#L37-L42

Reply:
Thanks @sachinpkale so when should `Store` be closed for remote store as in reference count decremented or explicitly closed?

Reply:
For local store we have `ShardStoreDeleter` as a listener that cleans up data, for remote we are deleting it as a part of close. Also when a shard relocates to another node, the shard engine I believe closes, would that mean we are removing data from the remote?

Reply:
Yes, we should not delete data on close. I think I have removed the `blobcontainer.delete()` as a part of some other commit which is not included in this PR. Let me add that.

I did not get the store leaking problem though. Are you talking about store or remoteStore?

Reply:
But internal engine is not aware of the remoteStore. Its scope is limited to IndexShard.